File: 프로그래머스/lv3/42895. N으로 표현/N으로 표현.py
import math

def solution(N, number):
    answer = 1
    results = [set([N])]
    if number == N:
        return answer

    while answer <= 8:
        new_set = set()
        
        for i in range(answer):
            for A in results[i]:
                for B in results[-i-1]:
                    new_set.add(A+B)
                    new_set.add(A*B)
                    new_set.add(A-B)
                    if B != 0:
                        new_set.add(A//B)
                        
        answer += 1                 
        new_set.add(int(str(N)*answer))    
        
        if number in new_set:
            return answer
        
        results.append(new_set)
    
    return -1

File: 프로그래머스/lv3/42895. N으로 표현/N으로 표현.java
import java.util.*;

class Solution {
    public int solution(int N, int number) {
        int answer = 0;
        String str = Integer.toString(N);
        
        ArrayList<HashSet<Integer>> dp = new ArrayList<>();
        System.out.println(dp.toString());
        
        for (int i=1; i<=8; i++) {
            HashSet<Integer> hs = new HashSet<>();
            
            String result = "";
            for (int k=0; k<i; k++) result += str;
            hs.add(Integer.valueOf(result));
            System.out.println(dp.toString());
            for (int j=1; j<=i/2; j++) {
                
                System.out.println(j-1 + " " + (i-j-1));
                HashSet<Integer> one = dp.get(j-1);
                HashSet<Integer> two = dp.get(i-j-1);
                
                for (int o : one) {
                    for (int t : two) {
                        hs.add(o+t);
                        hs.add(o-t);
                        hs.add(o*t);
                        if (t != 0) hs.add(o/t);
                        
                        hs.add(t+o);
                        hs.add(t-o);
                        hs.add(t*o);
                        if (o != 0) hs.add(t/o);
                    }
                }
            }
            
            if (hs.contains(number)) return i;
            dp.add(hs);
        }
        
        return -1;
    }
}